File: AnotherMe/lib/types/trace-event.ts
/**
 * Trace Events - Process event trace visualization.
 * 
 * Instead of only showing progress/step, this exposes internal workflow steps
 * as frontend-readable trace events. For problem video workflow, this shows:
 * - What knowledge points were identified
 * - What weaknesses were discovered
 * - What Manim errors were fixed
 * 
 * For course generation, this shows:
 * - Research phase results
 * - Outline generation details
 * - Scene generation progress per scene
 * - Quality gate results
 */

export type TraceEventType =
  | 'workflow_started'
  | 'workflow_step_started'
  | 'workflow_step_completed'
  | 'workflow_step_failed'
  | 'workflow_completed'
  | 'knowledge_identified'
  | 'weakness_discovered'
  | 'manim_error_fixed'
  | 'tts_generated'
  | 'video_rendered'
  | 'quality_gate_passed'
  | 'quality_gate_failed'
  | 'retry_attempted'
  | 'artifact_uploaded'
  | 'learner_profile_loaded'
  | 'adaptive_plan_generated';

export interface TraceEvent {
  /** Unique trace ID */
  id: string;
  /** Event type */
  type: TraceEventType;
  /** Timestamp */
  timestamp: number;
  /** Workflow/job ID this trace belongs to */
  jobId: string;
  /** Step name in the workflow */
  step: string;
  /** Duration of this step in milliseconds */
  durationMs: number | null;
  /** Status of this step */
  status: 'running' | 'completed' | 'failed' | 'skipped';
  /** Trace-specific payload */
  payload: TracePayload;
  /** Human-readable message */
  message: string;
  /** Severity level */
  severity: 'info' | 'warning' | 'error' | 'success';
}

export type TracePayload =
  | WorkflowStartedPayload
  | WorkflowStepPayload
  | KnowledgeIdentifiedPayload
  | WeaknessDiscoveredPayload
  | ManimErrorFixedPayload
  | TTSGeneratedPayload
  | VideoRenderedPayload
  | QualityGatePayload
  | RetryAttemptedPayload
  | ArtifactUploadedPayload
  | LearnerProfileLoadedPayload
  | AdaptivePlanPayload;

export interface WorkflowStartedPayload {
  type: 'workflow_started';
  workflowName: string;
  totalSteps: number;
  inputSummary: Record<string, unknown>;
}

export interface WorkflowStepPayload {
  type: 'workflow_step';
  stepIndex: number;
  totalSteps: number;
  details: Record<string, unknown>;
}

export interface KnowledgeIdentifiedPayload {
  type: 'knowledge_identified';
  knowledgePoints: string[];
  confidence: number;
  source: 'vision' | 'text' | 'inferred';
}

export interface WeaknessDiscoveredPayload {
  type: 'weakness_discovered';
  knowledgePoint: string;
  masteryScore: number;
  commonMistakes: string[];
  suggestedRemediation: string;
}

export interface ManimErrorFixedPayload {
  type: 'manim_error_fixed';
  errorType: string;
  errorMessage: string;
  fixApplied: string;
  retryCount: number;
}

export interface TTSGeneratedPayload {
  type: 'tts_generated';
  textLength: number;
  audioDurationSeconds: number;
  provider: string;
  voice: string;
}

export interface VideoRenderedPayload {
  type: 'video_rendered';
  resolution: string;
  fps: number;
  durationSeconds: number;
  fileSizeBytes: number;
}

export interface QualityGatePayload {
  type: 'quality_gate';
  checkName: string;
  passed: boolean;
  score: number;
  details: string;
}

export interface RetryAttemptedPayload {
  type: 'retry_attempted';
  stepName: string;
  attemptNumber: number;
  maxAttempts: number;
  reason: string;
}

export interface ArtifactUploadedPayload {
  type: 'artifact_uploaded';
  artifactType: string;
  artifactUrl: string;
  fileSizeBytes: number;
}

export interface LearnerProfileLoadedPayload {
  type: 'learner_profile_loaded';
  userId: string;
  weakSubjects: string[];
  weakKnowledgePoints: string[];
  abilityScores: Array<{ metric: string; value: number }>;
}

export interface AdaptivePlanPayload {
  type: 'adaptive_plan';
  mode: 'remedial' | 'standard' | 'advanced';
  weakKnowledgePoints: string[];
  ttsProfile: Record<string, string>;
  visualProfile: Record<string, string | boolean>;
}

/**
 * Creates a new trace event.
 */
export function createTraceEvent(
  jobId: string,
  type: TraceEventType,
  step: string,
  payload: TracePayload,
  message: string,
  severity: 'info' | 'warning' | 'error' | 'success' = 'info',
  status: 'running' | 'completed' | 'failed' | 'skipped' = 'running',
): TraceEvent {
  return {
    id: `trace-${Date.now()}-${Math.random().toString(36).slice(2, 10)}`,
    type,
    timestamp: Date.now(),
    jobId,
    step,
    durationMs: null,
    status,
    payload,
    message,
    severity,
  };
}

/**
 * Creates a workflow started trace event.
 */
export function createWorkflowStartedTrace(
  jobId: string,
  workflowName: string,
  totalSteps: number,
  inputSummary: Record<string, unknown> = {},
): TraceEvent {
  return createTraceEvent(
    jobId,
    'workflow_started',
    'init',
    {
      type: 'workflow_started',
      workflowName,
      totalSteps,
      inputSummary,
    },
    `Workflow "${workflowName}" started with ${totalSteps} steps`,
    'info',
    'running',
  );
}

/**
 * Creates a knowledge identified trace event.
 */
export function createKnowledgeIdentifiedTrace(
  jobId: string,
  step: string,
  knowledgePoints: string[],
  confidence: number,
  source: 'vision' | 'text' | 'inferred' = 'vision',
): TraceEvent {
  return createTraceEvent(
    jobId,
    'knowledge_identified',
    step,
    {
      type: 'knowledge_identified',
      knowledgePoints,
      confidence,
      source,
    },
    `Identified ${knowledgePoints.length} knowledge point(s): ${knowledgePoints.join(', ')}`,
    'success',
    'completed',
  );
}

/**
 * Creates a weakness discovered trace event.
 */
export function createWeaknessDiscoveredTrace(
  jobId: string,
  step: string,
  knowledgePoint: string,
  masteryScore: number,
  commonMistakes: string[],
  suggestedRemediation: string,
): TraceEvent {
  return createTraceEvent(
    jobId,
    'weakness_discovered',
    step,
    {
      type: 'weakness_discovered',
      knowledgePoint,
      masteryScore,
      commonMistakes,
      suggestedRemediation,
    },
    `Discovered weakness in "${knowledgePoint}" (mastery: ${Math.round(masteryScore * 100)}%)`,
    'warning',
    'completed',
  );
}

/**
 * Creates a Manim error fixed trace event.
 */
export function createManimErrorFixedTrace(
  jobId: string,
  step: string,
  errorType: string,
  errorMessage: string,
  fixApplied: string,
  retryCount: number,
): TraceEvent {
  return createTraceEvent(
    jobId,
    'manim_error_fixed',
    step,
    {
      type: 'manim_error_fixed',
      errorType,
      errorMessage,
      fixApplied,
      retryCount,
    },
    `Fixed Manim error: ${errorType} (${retryCount} retries)`,
    'warning',
    'completed',
  );
}

/**
 * Groups trace events by step for display.
 */
export function groupTracesByStep(traces: TraceEvent[]): Map<string, TraceEvent[]> {
  const grouped = new Map<string, TraceEvent[]>();
  
  for (const trace of traces) {
    const existing = grouped.get(trace.step) || [];
    existing.push(trace);
    grouped.set(trace.step, existing);
  }
  
  return grouped;
}

/**
 * Gets the current workflow status from traces.
 */
export function getWorkflowStatus(traces: TraceEvent[]): {
  status: 'not_started' | 'running' | 'completed' | 'failed';
  currentStep: string | null;
  progress: number;
  totalSteps: number;
  completedSteps: number;
} {
  if (traces.length === 0) {
    return { status: 'not_started', currentStep: null, progress: 0, totalSteps: 0, completedSteps: 0 };
  }
  
  const startedEvent = traces.find((t) => t.type === 'workflow_started');
  const totalSteps = (startedEvent?.payload as WorkflowStartedPayload)?.totalSteps ?? 0;
  
  const completedSteps = traces.filter((t) => t.status === 'completed' && t.type === 'workflow_step_completed').length;
  const failedSteps = traces.filter((t) => t.status === 'failed').length;
  
  const hasCompleted = traces.some((t) => t.type === 'workflow_completed');
  const hasFailed = traces.some((t) => t.type === 'workflow_step_failed' && t.severity === 'error');
  
  const lastRunning = [...traces].reverse().find((t) => t.status === 'running');
  
  return {
    status: hasCompleted ? 'completed' : hasFailed ? 'failed' : 'running',
    currentStep: lastRunning?.step ?? null,
    progress: totalSteps > 0 ? Math.round((completedSteps / totalSteps) * 100) : 0,
    totalSteps,
    completedSteps,
  };
}

/**
 * Gets a summary of knowledge points identified across all traces.
 */
export function getKnowledgePointSummary(traces: TraceEvent[]): {
  identified: string[];
  weaknesses: Array<{ point: string; mastery: number }>;
  errorsFixed: number;
} {
  const identified = new Set<string>();
  const weaknesses: Array<{ point: string; mastery: number }> = [];
  let errorsFixed = 0;
  
  for (const trace of traces) {
    if (trace.type === 'knowledge_identified') {
      const payload = trace.payload as KnowledgeIdentifiedPayload;
      for (const kp of payload.knowledgePoints) {
        identified.add(kp);
      }
    } else if (trace.type === 'weakness_discovered') {
      const payload = trace.payload as WeaknessDiscoveredPayload;
      weaknesses.push({ point: payload.knowledgePoint, mastery: payload.masteryScore });
    } else if (trace.type === 'manim_error_fixed') {
      errorsFixed++;
    }
  }
  
  return {
    identified: Array.from(identified),
    weaknesses,
    errorsFixed,
  };
}

File: AnotherMe/anotherme2_engine/agents/foundation/trace_event.py
"""
Trace Events - Process event trace visualization.

Instead of only showing progress/step, this exposes internal workflow steps
as frontend-readable trace events. For problem video workflow, this shows:
- What knowledge points were identified
- What weaknesses were discovered
- What Manim errors were fixed

For course generation, this shows:
- Research phase results
- Outline generation details
- Scene generation progress per scene
- Quality gate results
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional
from datetime import datetime
import uuid


VALID_TRACE_EVENT_TYPES = {
    "workflow_started",
    "workflow_step_started",
    "workflow_step_completed",
    "workflow_step_failed",
    "workflow_completed",
    "knowledge_identified",
    "weakness_discovered",
    "manim_error_fixed",
    "tts_generated",
    "video_rendered",
    "quality_gate_passed",
    "quality_gate_failed",
    "retry_attempted",
    "artifact_uploaded",
    "learner_profile_loaded",
    "adaptive_plan_generated",
}


@dataclass
class TraceEvent:
    """
    A trace event that exposes internal workflow steps for frontend visualization.
    
    Instead of just showing progress/step, trace events provide rich information
    about what happened during each step: knowledge points identified, weaknesses
    discovered, errors fixed, etc.
    """
    id: str = field(default_factory=lambda: str(uuid.uuid4())[:12])
    type: str = ""
    timestamp: Optional[datetime] = None
    job_id: str = ""
    step: str = ""
    duration_ms: Optional[int] = None
    status: str = "running"
    payload: Dict[str, Any] = field(default_factory=dict)
    message: str = ""
    severity: str = "info"

    def __post_init__(self):
        if self.timestamp is None:
            self.timestamp = datetime.utcnow()

    def to_dict(self) -> Dict[str, Any]:
        """Serialize to dictionary."""
        return {
            "id": self.id,
            "type": self.type,
            "timestamp": self.timestamp.isoformat() if self.timestamp else None,
            "job_id": self.job_id,
            "step": self.step,
            "duration_ms": self.duration_ms,
            "status": self.status,
            "payload": self.payload,
            "message": self.message,
            "severity": self.severity,
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "TraceEvent":
        """Deserialize from dictionary."""
        timestamp = None
        if data.get("timestamp"):
            try:
                timestamp = datetime.fromisoformat(data["timestamp"])
            except (ValueError, TypeError):
                pass
        
        return cls(
            id=data.get("id", str(uuid.uuid4())[:12]),
            type=data.get("type", ""),
            timestamp=timestamp,
            job_id=data.get("job_id", ""),
            step=data.get("step", ""),
            duration_ms=data.get("duration_ms"),
            status=data.get("status", "running"),
            payload=data.get("payload", {}),
            message=data.get("message", ""),
            severity=data.get("severity", "info"),
        )


class TraceEventEmitter:
    """
    Emits trace events for workflow visualization.
    
    This bridges the gap between backend workflow execution and frontend
    progress visualization, providing rich step-level details.
    """

    def __init__(self, job_id: str):
        self.job_id = job_id
        self.events: List[TraceEvent] = []
        self._step_start_times: Dict[str, datetime] = {}

    def emit(self, event: TraceEvent) -> TraceEvent:
        """Emit a trace event."""
        if not event.job_id:
            event.job_id = self.job_id
        if event.timestamp is None:
            event.timestamp = datetime.utcnow()
        
        self.events.append(event)
        return event

    def emit_workflow_started(self, total_steps: int = 0, message: str = "") -> TraceEvent:
        """Emit a workflow started trace event."""
        return self.emit(TraceEvent(
            type="workflow_started",
            step="workflow",
            status="running",
            message=message or "Workflow started",
            severity="info",
            payload={
                "type": "workflow_started",
                "total_steps": total_steps,
            },
        ))

    def emit_workflow_completed(self, message: str = "", payload: Optional[Dict[str, Any]] = None) -> TraceEvent:
        """Emit a workflow completed trace event."""
        return self.emit(TraceEvent(
            type="workflow_completed",
            step="workflow",
            status="completed",
            message=message or "Workflow completed",
            severity="success",
            payload=payload or {},
        ))

    def start_step(self, step: str, message: str = "") -> TraceEvent:
        """Mark the start of a workflow step."""
        self._step_start_times[step] = datetime.utcnow()
        
        return self.emit(TraceEvent(
            type="workflow_step_started",
            step=step,
            status="running",
            message=message or f"Step '{step}' started",
            severity="info",
        ))

    def complete_step(
        self,
        step: str,
        message: str = "",
        payload: Optional[Dict[str, Any]] = None,
    ) -> TraceEvent:
        """Mark a workflow step as completed."""
        duration_ms = None
        if step in self._step_start_times:
            delta = datetime.utcnow() - self._step_start_times[step]
            duration_ms = int(delta.total_seconds() * 1000)
            del self._step_start_times[step]
        
        return self.emit(TraceEvent(
            type="workflow_step_completed",
            step=step,
            status="completed",
            duration_ms=duration_ms,
            message=message or f"Step '{step}' completed",
            severity="success",
            payload=payload or {},
        ))

    def fail_step(
        self,
        step: str,
        error_message: str,
        payload: Optional[Dict[str, Any]] = None,
    ) -> TraceEvent:
        """Mark a workflow step as failed."""
        duration_ms = None
        if step in self._step_start_times:
            delta = datetime.utcnow() - self._step_start_times[step]
            duration_ms = int(delta.total_seconds() * 1000)
            del self._step_start_times[step]
        
        return self.emit(TraceEvent(
            type="workflow_step_failed",
            step=step,
            status="failed",
            duration_ms=duration_ms,
            message=error_message,
            severity="error",
            payload=payload or {},
        ))

    def emit_knowledge_identified(
        self,
        step: str,
        knowledge_points: List[str],
        confidence: float,
        source: str = "vision",
    ) -> TraceEvent:
        """Emit a knowledge identified trace event."""
        return self.emit(TraceEvent(
            type="knowledge_identified",
            step=step,
            status="completed",
            message=f"Identified {len(knowledge_points)} knowledge point(s): {', '.join(knowledge_points)}",
            severity="success",
            payload={
                "type": "knowledge_identified",
                "knowledge_points": knowledge_points,
                "confidence": confidence,
                "source": source,
            },
        ))

    def emit_weakness_discovered(
        self,
        step: str,
        knowledge_point: str,
        mastery_score: float,
        common_mistakes: List[str],
        suggested_remediation: str,
    ) -> TraceEvent:
        """Emit a weakness discovered trace event."""
        return self.emit(TraceEvent(
            type="weakness_discovered",
            step=step,
            status="completed",
            message=f'Discovered weakness in "{knowledge_point}" (mastery: {round(mastery_score * 100)}%)',
            severity="warning",
            payload={
                "type": "weakness_discovered",
                "knowledge_point": knowledge_point,
                "mastery_score": mastery_score,
                "common_mistakes": common_mistakes,
                "suggested_remediation": suggested_remediation,
            },
        ))

    def emit_manim_error_fixed(
        self,
        step: str,
        error_type: str,
        error_message: str,
        fix_applied: str,
        retry_count: int,
    ) -> TraceEvent:
        """Emit a Manim error fixed trace event."""
        return self.emit(TraceEvent(
            type="manim_error_fixed",
            step=step,
            status="completed",
            message=f"Fixed Manim error: {error_type} ({retry_count} retries)",
            severity="warning",
            payload={
                "type": "manim_error_fixed",
                "error_type": error_type,
                "error_message": error_message,
                "fix_applied": fix_applied,
                "retry_count": retry_count,
            },
        ))

    def emit_learner_profile_loaded(
        self,
        step: str,
        user_id: str,
        weak_subjects: List[str],
        weak_knowledge_points: List[str],
        ability_scores: List[Dict[str, Any]],
    ) -> TraceEvent:
        """Emit a learner profile loaded trace event."""
        return self.emit(TraceEvent(
            type="learner_profile_loaded",
            step=step,
            status="completed",
            message=f"Loaded learner profile for user {user_id}",
            severity="info",
            payload={
                "type": "learner_profile_loaded",
                "user_id": user_id,
                "weak_subjects": weak_subjects,
                "weak_knowledge_points": weak_knowledge_points,
                "ability_scores": ability_scores,
            },
        ))

    def emit_adaptive_plan(
        self,
        step: str,
        mode: str,
        weak_knowledge_points: List[str],
        tts_profile: Dict[str, str],
        visual_profile: Dict[str, Any],
    ) -> TraceEvent:
        """Emit an adaptive plan trace event."""
        return self.emit(TraceEvent(
            type="adaptive_plan_generated",
            step=step,
            status="completed",
            message=f"Generated adaptive plan: mode={mode}",
            severity="info",
            payload={
                "type": "adaptive_plan",
                "mode": mode,
                "weak_knowledge_points": weak_knowledge_points,
                "tts_profile": tts_profile,
                "visual_profile": visual_profile,
            },
        ))

    def emit_retry_attempted(
        self,
        step: str,
        attempt_number: int,
        max_attempts: int,
        reason: str,
    ) -> TraceEvent:
        """Emit a retry attempted trace event."""
        return self.emit(TraceEvent(
            type="retry_attempted",
            step=step,
            status="running",
            message=f"Retry attempt {attempt_number}/{max_attempts}: {reason}",
            severity="warning",
            payload={
                "type": "retry_attempted",
                "step_name": step,
                "attempt_number": attempt_number,
                "max_attempts": max_attempts,
                "reason": reason,
            },
        ))

    def get_events_by_step(self, step: str) -> List[TraceEvent]:
        """Get all events for a specific step."""
        return [e for e in self.events if e.step == step]

    def get_events_by_type(self, event_type: str) -> List[TraceEvent]:
        """Get all events of a specific type."""
        return [e for e in self.events if e.type == event_type]

    def get_knowledge_summary(self) -> Dict[str, Any]:
        """Get a summary of knowledge points and weaknesses from traces."""
        identified = set()
        weaknesses = []
        errors_fixed = 0
        
        for event in self.events:
            if event.type == "knowledge_identified":
                for kp in event.payload.get("knowledge_points", []):
                    identified.add(kp)
            elif event.type == "weakness_discovered":
                weaknesses.append({
                    "point": event.payload.get("knowledge_point", ""),
                    "mastery": event.payload.get("mastery_score", 0),
                })
            elif event.type == "manim_error_fixed":
                errors_fixed += 1
        
        return {
            "identified": sorted(identified),
            "weaknesses": weaknesses,
            "errors_fixed": errors_fixed,
        }

    def get_workflow_status(self) -> Dict[str, Any]:
        """Get the current workflow status from traces."""
        if not self.events:
            return {
                "status": "not_started",
                "current_step": None,
                "progress": 0,
                "total_steps": 0,
                "completed_steps": 0,
            }
        
        started_event = next(
            (e for e in self.events if e.type == "workflow_started"),
            None,
        )
        total_steps = started_event.payload.get("total_steps", 0) if started_event else 0
        
        completed_steps = sum(
            1 for e in self.events
            if e.status == "completed" and e.type == "workflow_step_completed"
        )
        
        has_completed = any(e.type == "workflow_completed" for e in self.events)
        has_failed = any(
            e.type == "workflow_step_failed" and e.severity == "error"
            for e in self.events
        )
        
        last_running = next(
            (e for e in reversed(self.events) if e.status == "running"),
            None,
        )
        
        progress = round((completed_steps / total_steps * 100)) if total_steps > 0 else 0
        
        return {
            "status": "completed" if has_completed else ("failed" if has_failed else "running"),
            "current_step": last_running.step if last_running else None,
            "progress": progress,
            "total_steps": total_steps,
            "completed_steps": completed_steps,
        }

    def to_event_list(self) -> List[Dict[str, Any]]:
        """Convert all events to dictionary list."""
        return [e.to_dict() for e in self.events]
